File: 3.无重复字符的最长子串.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>
using namespace std;
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */
/**
 * 出现的问题：
 * 1.每次循环后未将 unordered_map 进行清空
 * 解决方法：使用 isExit.clear() 清空 map 中的键值对
 * 2.当 " " 会导致出现问题，未进入循环，直接 return 0，但是这种情况需要 return 1
 * 解决方法：进行判断 length <= 1 的时候直接 return length
*/
/**
 * 暴力循环,确实低效
*/
/**
 * Your runtime beats 5.02 % of cpp submissions
 * Your memory usage beats 5.01 % of cpp submissions (158.6 MB)
*/
// @lc code=start
class MySolution {
public:
    int lengthOfLongestSubstring(string s) {
        int length = s.length();
        int maxSubStrlength = 0;
        unordered_map<char,bool> isExit;
        if (length <= 1)
        {
            return length;
        }
        
        for (int i = 0; i < length -1; i++)
        {
            string subStr = "";
            int subStrLength = 0;
            for (int j = i; j < length; j++)
            {
                if (!isExit[s[j]])
                {
                    subStrLength++;
                    subStr.push_back(s[j]);
                    isExit[s[j]] = true;
                }
                else
                {
                    break;
                }
            }
            printf("subStrlength : %d",subStrLength);
            isExit.clear();
            maxSubStrlength = maxSubStrlength > subStrLength ? maxSubStrlength : subStrLength;
        }
        return maxSubStrlength;
    }
};
/**
 * 题解的方法名字叫做：滑动窗口
 * 有点懂这个想法了，在上一次遍历中，将左边界左移，但是右边界不移动的话，此范围内一定是无重复的
 * 此时，再将右边界不断的右移扩展右边界，将边界扩大，直到出现重复的元素。而且此时维护已出现元素
 * 不需要进行清空，只需要一直维护即可
 * 刚开始的左右边界位置也很🐂
 * 左边界的初始位置为 0 ，右边界的初始位置为 -1,表示并未开始移动右边界，即左右边界内的元素个数为0；
 * 当左右边界相等时，表示此中只有一个元素。
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int right = -1;
        int ans = 0;
        unordered_set<char> isExit;
        for (int i = 0; i < n; i++)
        {
            if (i != 0)
            {
                // i != 0 表示左边界已经移动了，此时需要将左边界上一个位置的元素从已出现的 set 中删除
                isExit.erase(s[i - 1]);
            }
            // right + 1 < n，表示，右边界移动到下一个位置
            while (right + 1 < n && !isExit.count(s[right + 1]))
            {
                isExit.insert(s[right + 1]);
                right++;
            }
            // right - i + 1 未当前左右边界内的元素个数
            ans = max(ans,right - i + 1);
        }
        return ans;
    }
};
// @lc code=end

File: 3.无重复字符的最长子串.java

import java.util.HashSet;

/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */
/**
 * 再战
 * 滑动窗口
 * 又不会了
 */
// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        HashSet<Character> appearedChar = new HashSet<Character>();
        if (n == 0) {
            return 0;
        }
        // // int left = -1, right = 0, notRepeating = 0, result = 0;
        // // 感觉不用维护 notRepeating 因为 left 到 right 中维护的就是不重复的。
        // int left = -1, right = 0, result = 0;
        // while (right < n - 1) {
        // if (!appearedChar.contains(s.charAt(right + 1))) {
        // notRepeating += 1;
        // }
        // appearedChar.remove(s.charAt(++left));
        // notRepeating--;
        // result = Math.max(notRepeating, result);
        // }
        // return result;
        // 其实就是维护一个窗口，在以 left 为起点下，最大长度的无重复字符串，对应的 right 下标，当 left 移动的时候，不断向右移动 right
        // ，更新 right 获得新的 left 下最大最大的长度的无重复字符串
        int right = -1, result = 0;
        for (int left = 0; left < n; left++) {
            // 当 left = 0,&& right = -1 的时候表示没有开始
            if (left != 0) {
                // 移去最左侧的元素
                appearedChar.remove(s.charAt(left - 1));
            }
            // 如果 right + 1，即右边界还可以扩展没有达到最右边，并且 right 没有出现的话。
            while (right + 1 < n && !appearedChar.contains(s.charAt(right + 1))) {
                appearedChar.add(s.charAt(right + 1));
                right++;
            }
            result = Math.max(result, right - left + 1);
        }
        return result;
    }
}
// @lc code=end
